### PR TITLE
EXP-2499 Support for Web Libraries

### PR DESCRIPTION
**Pull request by oliver-richers**

### CODE REVIEW

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java [hunk 1]
```diff
@@ -1195,6 +1195,12 @@ public interface CoreI18n extends EmfI18n {
 		.de("Version");
 	I18n0 WATCHED_DAEMONS = new I18n0("Watched Daemons")//
 		.de("Überwachte Daemons");
+	I18n0 WEB_SERVICE_PATH_IS_ILLEGAL = new I18n0("Web service path is illegal.")//
+		.de("Webservice-Pfad ist ungültig.");
+	I18n0 WEB_SERVICE_UUID_IS_ILLEGAL = new I18n0("Web service UUID is illegal.")//
+		.de("Webservice UUID ist ungültig.");
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:1201]:
```suggestion
		.de("Webservice-UUID ist ungültig.");
```

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java [hunk 2]
```diff
@@ -1195,6 +1195,12 @@ public interface CoreI18n extends EmfI18n {
...
+	I18n0 WEB_SERVICE_UUID_IS_MISSING = new I18n0("Web service UUID is missing.")//
+		.de("Webservice UUID fehlt.");
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:1203]:
```suggestion
		.de("Webservice-UUID fehlt.");
```

File: platform-core-module/src/main/java/com/softicar/platform/core/module/web/service/WebServicePath.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	 *            the path (never <i>null</i>)
+	 * @return the parsed {@link WebServicePath}
+	 */
+	public static WebServicePath parseOrThrow(String path) {
+
+		return parse(path).orElseThrow(() -> new SofticarUserException(CoreI18n.WEB_SERVICE_PATH_IS_ILLEGAL));
+	}
+
+	/**
+	 * Returns the service identifier (usually a UUID).
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/web/service/WebServicePath.java:60]: We should add an example here (i.e. what this method returns for an exemplary path).

File: platform-core-module/src/main/java/com/softicar/platform/core/module/web/service/WebServicePath.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	 *
+	 * @return the service identifier (never <i>null</i>)
+	 */
+	public String getServiceIdentifier() {
+
+		return serviceIdentifier;
+	}
+
+	/**
+	 * Returns the service resource path, that is, the part after the service
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/web/service/WebServicePath.java:70]: We should add an example here (i.e. what this method returns for an exemplary path).

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java (hunk 1 above)
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:1201]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java (hunk 2 above)
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:1203]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/web/service/WebServicePath.java (hunk 4 above)
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/web/service/WebServicePath.java:70]: added

File: platform-core-module/src/test/java/com/softicar/platform/core/module/web/service/WebServicePathTest.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.softicar.platform.common.testing.Asserts;
+import org.junit.Test;
+
+public class WebServicePathTest extends Asserts {
+
+	@Test
+	public void testParseWithValidPath() {
+
+		assertEquals("[,]", assertOne(WebServicePath.parse("/service/")).toString());
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/web/service/WebServicePathTest.java:11]: ok, changed

File: platform-core-module/src/test/java/com/softicar/platform/core/module/web/service/WebServicePathTest.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+		assertEquals("[0815,]", assertOne(WebServicePath.parse("/service/0815")).toString());
+		assertEquals("[0815,/]", assertOne(WebServicePath.parse("/service/0815/")).toString());
+		assertEquals("[0815,/foo/bar]", assertOne(WebServicePath.parse("/service/0815/foo/bar")).toString());
+	}
+
+	@Test
+	public void testParseWithIllegalPath() {
+
+		assertEmpty(WebServicePath.parse(null));
+		assertEmpty(WebServicePath.parse(""));
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/web/service/WebServicePathTest.java:22]: added

File: platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibrary.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * @author Oliver Richers
+ */
+public class WebLibrary implements IWebServiceFactory {
+
+	private final String packagePath;
+
+	/**
+	 * The Java package path to the content of the web library.
+	 * <p>
+	 * It is assumed that the web library is included into the Java class path
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibrary.java:20]: changed to `is part of the Java class path`

File: platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibrary.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	 * as a Jar file.
+	 * <p>
+	 * <b>Warning:</b> Be sure to supply a path as deep as possible to not
+	 * expose resources that should remain confidential. For example, supplying
+	 * the path <code>"/"</code> would provide access to the whole code base.
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibrary.java:25]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibraryService.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.OutputStream;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * The {@link IWebService} implementation used by {@link WebLibrary}.
+ *
+ * @author Oliver Richers
+ */
+class WebLibraryService implements IWebService {
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibraryService.java:23]: not sure if I like this better, but it's ok, changed

File: platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibraryService.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+				}
+			} else {
+				throw new IllegalArgumentException("Missing resource: " + resourcePath);
+			}
+		} catch (IOException exception) {
+			throw new SofticarIOException(exception);
+		}
+	}
+
+	private void copySafely(InputStream inputStream, OutputStream outputStream) {
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibraryService.java:50]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibraryService.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+			new StreamCopy(inputStream, outputStream).copyAndClose();
+		} catch (StreamCopyOutputException exception) {
+			DevNull.swallow(exception);
+			Log.error("Failed to transfer resource to HTTP client. The client probably closed the connection prematurely.");
+		}
+	}
+
+	private MimeType determineContentType(String path) {
+
+		return MimeType.getByFilenameOrDefault(path);
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/web/library/WebLibraryService.java:62]: I don't see a need for a warning. Added a setter method for the mime-type determiner.

**Pull request merged**